src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.InlineContext.cs

```diff
@@ -0,0 +1,582 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.InlineMethod
+{
+    internal abstract partial class AbstractInlineMethodRefactoringProvider<TInvocationNode, TExpression, TArgumentSyntax, TMethodDeclarationSyntax, TIdentifierName>
+        where TInvocationNode : SyntaxNode
+        where TExpression : SyntaxNode
+        where TArgumentSyntax : SyntaxNode
+        where TMethodDeclarationSyntax : SyntaxNode
+        where TIdentifierName : SyntaxNode
+    {
+        private class InlineMethodContext
+        {
+            /// <summary>
+            /// Statements should be inserted before the <see cref="StatementContainingCallee"/>.
+            /// </summary>
+            public ImmutableArray<SyntaxNode> StatementsToInsertBeforeCallee { get; }
+
+            /// <summary>
+            /// Statement containing the callee. All the generated declarations should be put before this node.
+            /// Example:
+            /// void Caller()
+            /// {
+            ///     var x = Callee(Foo());
+            /// }
+            /// void Callee(int i) { DoSomething(); }
+            /// LocalDeclarationSyntaxNode of x will the <see cref="StatementContainingCallee"/>.
+            /// And if there is any statements needs inserted, it needs to be inserted before this node.
+            /// </summary>
+            public SyntaxNode StatementContainingCallee { get; }
+
+            /// <summary>
+            /// Inline content for the callee method. It should replace <see cref="SyntaxNodeToReplace"/>.
+            /// </summary>
+            public SyntaxNode InlineSyntaxNode { get; }
+
+            /// <summary>
+            /// SyntaxNode needs to be replaced by <see cref="InlineSyntaxNode"/>
+            /// </summary>
+            public SyntaxNode SyntaxNodeToReplace { get; }
+
+            /// <summary>
+            /// Indicate is <see cref="InlineSyntaxNode"/> has Await Expression.
+            /// </summary>
+            public bool ContainsAwaitExpression { get; }
```

Comment:
will have to see your impl.  but in general if yuor'e just walking blindly down lookign for 'await' then this won't be correct.